packages/back-end/src/services/experiments.ts

```diff
@@ -1075,6 +1065,40 @@ export function postMetricApiPayloadIsValid(
   return {
     valid: true,
   };
+};
+
+/**
+ * While the `postMetricValidator` can detect the presence of values, it cannot figure out the correctness.
+ * @param payload
+ * @param datasource
+ */
+export function postMetricApiPayloadIsValid(
+  payload: z.infer<typeof postMetricValidator.bodySchema>,
+  datasource: Pick<DataSourceInterface, "type">
+): { valid: true } | { valid: false; error: string } {
+  if (!payload.sql && !payload.sqlBuilder && !payload.mixpanel) {
+    return {
+      valid: false,
+      error: "Must specify one of: sql, sqlBuilder, mixpanel",
+    };
+  }
+
+  const { mixpanel } = payload;
+
+  if (mixpanel && datasource.type !== "mixpanel") {
+    return {
+      valid: false,
+      error: "Mixpanel datasources must provide `mixpanel`",
+    };
+  }
+
+  return postOrPutMetricPayloadIsValid(payload);
+}
+
+export function putMetricApiPayloadIsValid(
+  payload: z.infer<typeof putMetricValidator.bodySchema>
+): { valid: true } | { valid: false; error: string } {
+  return postOrPutMetricPayloadIsValid(payload);
```

Comment:
I see  what you're doing here re: the abstraction. I think copying what's needed is the best approach rather than trying to DRY it up, because now the post validation is segmented across 2 separate functions.